File: packages/pigeon/platform_tests/alternate_language_test_plugin/android/src/test/java/com/example/alternate_language_test_plugin/NullFieldsTest.java
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.example.alternate_language_test_plugin;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;

import java.util.ArrayList;
import java.util.Arrays;
import org.junit.Test;

public class NullFieldsTest {
  @Test
  public void builderWithValues() {
    NullFields.NullFieldsSearchRequest request =
        new NullFields.NullFieldsSearchRequest.Builder()
            .setQuery("hello")
            .setIdentifier(1L)
            .build();

    NullFields.NullFieldsSearchReply reply =
        new NullFields.NullFieldsSearchReply.Builder()
            .setResult("result")
            .setError("error")
            .setIndices(Arrays.asList(1L, 2L, 3L))
            .setRequest(request)
            .setType(NullFields.NullFieldsSearchReplyType.SUCCESS)
            .build();

    assertEquals(reply.getResult(), "result");
    assertEquals(reply.getError(), "error");
    assertEquals(reply.getIndices(), Arrays.asList(1L, 2L, 3L));
    assertEquals(reply.getRequest().getQuery(), "hello");
    assertEquals(reply.getType(), NullFields.NullFieldsSearchReplyType.SUCCESS);
  }

  @Test
  public void builderRequestWithNulls() {
    NullFields.NullFieldsSearchRequest request =
        new NullFields.NullFieldsSearchRequest.Builder().setQuery(null).setIdentifier(1L).build();
  }

  @Test
  public void builderReplyWithNulls() {
    NullFields.NullFieldsSearchReply reply =
        new NullFields.NullFieldsSearchReply.Builder()
            .setResult(null)
            .setError(null)
            .setIndices(null)
            .setRequest(null)
            .setType(null)
            .build();

    assertNull(reply.getResult());
    assertNull(reply.getError());
    assertNull(reply.getIndices());
    assertNull(reply.getRequest());
    assertNull(reply.getType());
  }

  @Test
  public void requestFromMapWithValues() {
    ArrayList<Object> list = new ArrayList<Object>();
    list.add("hello");
    list.add(1L);
    NullFields.NullFieldsSearchRequest request = NullFields.NullFieldsSearchRequest.fromList(list);
    assertEquals(request.getQuery(), "hello");
  }

  @Test
  public void requestFromMapWithNulls() {
    ArrayList<Object> list = new ArrayList<Object>();
    list.add(null);
    list.add(1L);

    NullFields.NullFieldsSearchRequest request = NullFields.NullFieldsSearchRequest.fromList(list);
    assertNull(request.getQuery());
  }

  @Test
  public void replyFromMapWithValues() {
    NullFields.NullFieldsSearchRequest request =
        new NullFields.NullFieldsSearchRequest.Builder()
            .setQuery("hello")
            .setIdentifier(1L)
            .build();

    NullFields.NullFieldsSearchReply input =
        new NullFields.NullFieldsSearchReply.Builder()
            .setResult("result")
            .setError("error")
            .setIndices(Arrays.asList(1L, 2L, 3L))
            .setRequest(request)
            .setType(NullFields.NullFieldsSearchReplyType.SUCCESS)
            .build();

    NullFields.NullFieldsSearchReply reply =
        NullFields.NullFieldsSearchReply.fromList(input.toList());
    assertEquals(reply.getResult(), "result");
    assertEquals(reply.getError(), "error");
    assertEquals(reply.getIndices(), Arrays.asList(1L, 2L, 3L));
    assertEquals(reply.getRequest().getQuery(), "hello");
    assertEquals(reply.getType(), NullFields.NullFieldsSearchReplyType.SUCCESS);
  }

  @Test
  public void replyFromMapWithNulls() {
    ArrayList<Object> list = new ArrayList<Object>();

    list.add(null);
    list.add(null);
    list.add(null);
    list.add(null);
    list.add(null);

    NullFields.NullFieldsSearchReply reply = NullFields.NullFieldsSearchReply.fromList(list);
    assertNull(reply.getResult());
    assertNull(reply.getError());
    assertNull(reply.getIndices());
    assertNull(reply.getRequest());
    assertNull(reply.getType());
  }

  @Test
  public void requestToMapWithValues() {
    NullFields.NullFieldsSearchRequest request =
        new NullFields.NullFieldsSearchRequest.Builder()
            .setQuery("hello")
            .setIdentifier(1L)
            .build();

    ArrayList<Object> list = request.toList();
    assertEquals(list.get(0), "hello");
  }

  @Test
  public void requestToMapWithNulls() {
    NullFields.NullFieldsSearchRequest request =
        new NullFields.NullFieldsSearchRequest.Builder().setQuery(null).setIdentifier(1L).build();

    ArrayList<Object> list = request.toList();
    assertNull(list.get(0));
  }

  @Test
  public void replyToMapWithValues() {
    NullFields.NullFieldsSearchRequest request =
        new NullFields.NullFieldsSearchRequest.Builder()
            .setQuery("hello")
            .setIdentifier(1L)
            .build();

    NullFields.NullFieldsSearchReply reply =
        new NullFields.NullFieldsSearchReply.Builder()
            .setResult("result")
            .setError("error")
            .setIndices(Arrays.asList(1L, 2L, 3L))
            .setRequest(request)
            .setType(NullFields.NullFieldsSearchReplyType.SUCCESS)
            .build();

    ArrayList<Object> list = reply.toList();
    assertEquals(list.get(0), "result");
    assertEquals(list.get(1), "error");
    assertEquals(list.get(2), Arrays.asList(1L, 2L, 3L));
    assertEquals(list.get(3), reply.getRequest());
    assertEquals(list.get(4), NullFields.NullFieldsSearchReplyType.SUCCESS.ordinal());
  }

  @Test
  public void replyToMapWithNulls() {
    NullFields.NullFieldsSearchReply reply =
        new NullFields.NullFieldsSearchReply.Builder()
            .setResult(null)
            .setError(null)
            .setIndices(null)
            .setRequest(null)
            .setType(null)
            .build();

    ArrayList<Object> list = reply.toList();

    assertNull(list.get(0));
    assertNull(list.get(1));
    assertNull(list.get(2));
    assertNull(list.get(3));
    assertNull(list.get(4));
  }
}

File: packages/pigeon/platform_tests/alternate_language_test_plugin/example/ios/RunnerTests/NullFieldsTest.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

@import Flutter;
@import XCTest;

@import alternate_language_test_plugin;

#import "EchoMessenger.h"

///////////////////////////////////////////////////////////////////////////////////////////
@interface NullFieldsSearchRequest ()
+ (NullFieldsSearchRequest *)fromList:(NSArray<id> *)list;
- (NSArray<id> *)toList;
@end

///////////////////////////////////////////////////////////////////////////////////////////
@interface NullFieldsSearchReply ()
+ (NullFieldsSearchReply *)fromList:(NSArray<id> *)list;
- (NSArray<id> *)toList;
@end

///////////////////////////////////////////////////////////////////////////////////////////
@interface NullFieldsTest : XCTestCase
@end

///////////////////////////////////////////////////////////////////////////////////////////
@implementation NullFieldsTest

- (void)testMakeWithValues {
  NullFieldsSearchRequest *request = [NullFieldsSearchRequest makeWithQuery:@"hello" identifier:1];

  NullFieldsSearchReplyTypeBox *typeWrapper =
      [[NullFieldsSearchReplyTypeBox alloc] initWithValue:NullFieldsSearchReplyTypeSuccess];
  NullFieldsSearchReply *reply = [NullFieldsSearchReply makeWithResult:@"result"
                                                                 error:@"error"
                                                               indices:@[ @1, @2, @3 ]
                                                               request:request
                                                                  type:typeWrapper];

  NSArray<id> *indices = @[ @1, @2, @3 ];
  XCTAssertEqualObjects(@"result", reply.result);
  XCTAssertEqualObjects(@"error", reply.error);
  XCTAssertEqualObjects(indices, reply.indices);
  XCTAssertEqualObjects(@"hello", reply.request.query);
  XCTAssertEqual(typeWrapper.value, reply.type.value);
}

- (void)testMakeRequestWithNulls {
  NullFieldsSearchRequest *request = [NullFieldsSearchRequest makeWithQuery:nil identifier:1];
  XCTAssertNil(request.query);
}

- (void)testMakeReplyWithNulls {
  NullFieldsSearchReply *reply = [NullFieldsSearchReply makeWithResult:nil
                                                                 error:nil
                                                               indices:nil
                                                               request:nil
                                                                  type:nil];
  XCTAssertNil(reply.result);
  XCTAssertNil(reply.error);
  XCTAssertNil(reply.indices);
  XCTAssertNil(reply.request);
  XCTAssertNil(reply.type);
}

- (void)testRequestFromListWithValues {
  NSArray<id> *list = @[
    @"hello",
    @1,
  ];
  NullFieldsSearchRequest *request = [NullFieldsSearchRequest fromList:list];
  XCTAssertEqualObjects(@"hello", request.query);
}

- (void)testRequestFromListWithNulls {
  NSArray<id> *list = @[
    [NSNull null],
    @1,
  ];
  NullFieldsSearchRequest *request = [NullFieldsSearchRequest fromList:list];
  XCTAssertNil(request.query);
}

- (void)testReplyFromListWithValues {
  NullFieldsSearchReplyTypeBox *typeWrapper =
      [[NullFieldsSearchReplyTypeBox alloc] initWithValue:NullFieldsSearchReplyTypeSuccess];

  NSArray<id> *listRequest = @[
    @"hello",
    @1,
  ];
  NSArray<id> *listReply = @[
    @"result",
    @"error",
    @[ @1, @2, @3 ],
    [NullFieldsSearchRequest fromList:listRequest],
    typeWrapper,
  ];

  NSArray<id> *indices = @[ @1, @2, @3 ];
  NullFieldsSearchReply *reply = [NullFieldsSearchReply fromList:listReply];

  XCTAssertEqualObjects(@"result", reply.result);
  XCTAssertEqualObjects(@"error", reply.error);
  XCTAssertEqualObjects(indices, reply.indices);
  XCTAssertEqualObjects(@"hello", reply.request.query);
  XCTAssertEqual(NullFieldsSearchReplyTypeSuccess, reply.type.value);
}

- (void)testReplyFromListWithNulls {
  NSArray<id> *list = @[
    [NSNull null],
    [NSNull null],
    [NSNull null],
    [NSNull null],
    [NSNull null],
  ];
  NullFieldsSearchReply *reply = [NullFieldsSearchReply fromList:list];
  XCTAssertNil(reply.result);
  XCTAssertNil(reply.error);
  XCTAssertNil(reply.indices);
  XCTAssertNil(reply.request.query);
  XCTAssertNil(reply.type);
}

- (void)testRequestToListWithValuess {
  NullFieldsSearchRequest *request = [NullFieldsSearchRequest makeWithQuery:@"hello" identifier:1];
  NSArray<id> *list = [request toList];
  XCTAssertEqual(@"hello", list[0]);
}

- (void)testRequestToListWithNulls {
  NullFieldsSearchRequest *request = [NullFieldsSearchRequest makeWithQuery:nil identifier:1];
  NSArray<id> *list = [request toList];
  XCTAssertEqual([NSNull null], list[0]);
}

- (void)testReplyToListWithValuess {
  NullFieldsSearchReplyTypeBox *typeWrapper =
      [[NullFieldsSearchReplyTypeBox alloc] initWithValue:NullFieldsSearchReplyTypeSuccess];
  NullFieldsSearchReply *reply = [NullFieldsSearchReply
      makeWithResult:@"result"
               error:@"error"
             indices:@[ @1, @2, @3 ]
             request:[NullFieldsSearchRequest makeWithQuery:@"hello" identifier:1]
                type:typeWrapper];
  NSArray<id> *list = [reply toList];
  NSArray<id> *indices = @[ @1, @2, @3 ];
  XCTAssertEqualObjects(@"result", list[0]);
  XCTAssertEqualObjects(@"error", list[1]);
  XCTAssertEqualObjects(indices, list[2]);
  XCTAssertEqualObjects(@"hello", ((NullFieldsSearchRequest *)list[3]).query);
  NullFieldsSearchReplyTypeBox *output = list[4];

  XCTAssertEqual(typeWrapper.value, output.value);
}

- (void)testReplyToListWithNulls {
  NullFieldsSearchReply *reply = [NullFieldsSearchReply makeWithResult:nil
                                                                 error:nil
                                                               indices:nil
                                                               request:nil
                                                                  type:nil];
  NSArray<id> *list = [reply toList];
  XCTAssertEqual([NSNull null], list[0]);
  XCTAssertEqual([NSNull null], list[1]);
  XCTAssertEqual([NSNull null], list[2]);
  XCTAssertEqual([NSNull null], list[3]);
  XCTAssertEqual([NSNull null], list[4]);
}

@end
